### DISCUSSION TITLE
Expandable Text

### DISCUSSION THREAD

**Discussion by iamclaytonray** [chakra-ui/chakra-ui#5471] [Ideas]
`noOfLines={NUMBER}` is amazing but something that would be amazing-er would be to add an option to expand the text to it's complete state, adding a prop to modify what's rendered (ie: `Read more`, `See more`, etc)

**Comment by dz3n** [comment]
Is there workaround for this? How can I put a "read more" inside **noOfLines** element

**Comment by iamclaytonray** [comment]
This is what I have at the moment. I don't calculate whether the `Read more` button **_should_** show up so there are times where 100% of the text is visible and clicking the button doesn't expand anything (because there is nothing to expand). So that is one caveat with this code

You could add some changes, such as the `Show more` and `Show less` text. ie: `props.showMoreText`

```tsx
import { Box, Button } from '@chakra-ui/react';
import { memo, ReactNode, useState } from 'react';

interface Props {
  noOfLines?: number;
  children: ReactNode;
}

export const ExpandableText = memo(({ noOfLines = 4, children }: Props) => {
  const [expandedCount, setExpandedCount] = useState(noOfLines);

  const handleToggle = () =>
    setExpandedCount(expandedCount ? undefined : noOfLines);

  return (
    <Box
      display="inline-block"
      as="span"
    >
      <Box noOfLines={expandedCount}>{children}</Box>
      <Button
        size="sm"
        variant="link"
        fontWeight="bold"
        colorScheme="slate"
        textDecoration="underline"
        onClick={handleToggle}
      >
        {!expandedCount ? 'Show less' : 'Read more'}
      </Button>
    </Box>
  );
});
```

**Reply by iamclaytonray** [reply]
If you do use this code and figure out a way to calculate whether the expand button should appear or not, could you add that code block here?

**Reply by dz3n** [reply]
I think we can determinate this if compare **scrollHeight > clientHeight**, but if you click it twice (one time to expand and another to bring it back to normal) those values become equal and the read more button disappear. That's why I use isClicked state.

here is my version of your code.

```javascript
import type { BoxProps } from '@chakra-ui/react';
import { Box, Button, Text } from '@chakra-ui/react';
import React, { forwardRef, useState } from 'react';

interface Props extends BoxProps {
  children: React.ReactNode;
  noOfLines: number;
}

export const ExpandableText = forwardRef<HTMLDivElement, Props>(
  ({ children, noOfLines, ...rest }, ref) => {
    const [expandedCount, setExpandedCount] = useState<number | undefined>(
      noOfLines
    );
    const [isClicked, setIsClicked] = useState(false);
    const handleToggle = () => {
      setIsClicked(true);
      setExpandedCount(expandedCount ? undefined : noOfLines);
    };

    const inputRef = React.useRef<HTMLInputElement>(null);

    const isTextClamped =
      (inputRef.current?.scrollHeight as number) >
        (inputRef.current?.clientHeight as number) || isClicked;

    return (
      <Box ref={ref} {...rest}>
        <Box ref={inputRef} noOfLines={expandedCount}>
          {children}
        </Box>
        <Button
          display={isTextClamped ? 'block' : 'none'}
          size="sm"
          variant="link"
          onClick={handleToggle}
        >
          <Text>{!expandedCount ? 'Show less' : 'Read more'}</Text>
        </Button>
      </Box>
    );
  }
);

ExpandableText.displayName = 'ExpandableText';

```

**Reply by nazhenhuiyi** [reply]
very nice! thank you!

**Comment by savasgok** [comment]
This one works better.

import { BoxProps, Center } from '@chakra-ui/react';
import { Box, Button, Text } from '@chakra-ui/react';
import React, { forwardRef, useLayoutEffect, useState } from 'react';

interface Props extends BoxProps {
    children: React.ReactNode;
    noOfLines: number;
}

export const ExpandableText = forwardRef<HTMLDivElement, Props>(
    ({ children, noOfLines, ...rest }, ref) => {
        const [expandedCount, setExpandedCount] = useState<number | undefined>(
            noOfLines
        );
        const [isClicked, setIsClicked] = useState(false);
        const handleToggle = () => {
            setIsClicked(true);
            setExpandedCount(expandedCount ? undefined : noOfLines);
        };
        const [isTextClamped, setIsTextClamped]=useState(false);
        
        const inputRef = React.useRef<HTMLInputElement>(null);
        

       
          
        useLayoutEffect(() => {
            let textClamped=false
            if (inputRef.current && inputRef.current.clientHeight) {
                textClamped = (inputRef.current?.scrollHeight as number) >
                    (inputRef.current?.clientHeight as number) || isClicked;
                    setIsTextClamped(textClamped)
              
                    
            }

        }, [setIsTextClamped]);

      
       

        return (
            <Box ref={ref} {...rest}>
                <Box ref={inputRef} noOfLines={expandedCount}>
                    {children}
                </Box>
                <Center>
                    <Button
                        display={isTextClamped ? 'block' : 'none'}
                        size="sm"
                        variant="link"
                        onClick={handleToggle}
                        mt="2"
                    >
                        <Text>{!expandedCount ? 'Show less' : 'Show more'}</Text>
                    </Button></Center>
            </Box>
        );
    }
);

ExpandableText.displayName = 'ExpandableText';
